JSONRPC: distinguish not-found and unavailable from invalid params

malard opened the pull request:
## Description

JSON-RPC returns `InvalidParams` (`-32602`) both when parameters are genuinely malformed and when they are well formed but reference an object that does not exist. A client cannot tell a bad request from a stale reference.

`Player.Open` compounds this. Playback is posted asynchronously via `TMSG_MEDIA_PLAY`, so the method returns `ACK` before anything is attempted — it reports success even when the item cannot be reached, and nothing downstream can be surfaced to the caller.

This adds two statuses inside the implementation-defined range:

| Status | Code | Meaning |
| --- | --- | --- |
| `NotFound` | `-32098` | the referenced object does not exist |
| `Unavailable` | `-32097` | the reference is valid, the content cannot be reached |

The distinction matters because the remedies are opposite. `NotFound` means the reference should be discarded. `Unavailable` means it should be kept and retried once the storage is back. Collapsing both into `InvalidParams` means an automation that bookmarked a movie cannot tell "this film was removed" from "your NAS is offline".

This supersedes the original proposal in this PR of a single `InvalidObject = -32101`, which @ksooo correctly noted was named and worded for the database case only. A general `NotFound` covers the database, the VFS, PVR and add-ons without needing a near-duplicate code later.

## Changes

- `JSONRPCUtils.h` — adds `NotFound` and `Unavailable`. The reserved range comment is corrected from `-32099..-32000` to `-32100..-32000` so that it matches the codes already in use; `FailedToExecute = -32100` therefore does not have to move.
- `JSONRPC.cpp` — response building for both statuses.
- `VideoLibrary.cpp` — 17 sites return `NotFound` where a library object is absent.
- `AudioLibrary.cpp` — the 8 equivalent artist/album/song sites, so music and video behave the same. The remaining `InvalidParams` in both files are genuine parameter faults (`GetXspFiltering`, `ParseSorting`) and are unchanged.
- `PlayerOperations.cpp` — `Player.Open` verifies a single addressable item before posting playback and returns `Unavailable` if it cannot be reached. Folders, plugins, UPnP, internet streams and Bluray paths are left to the player as before. The directory cache is bypassed so a cached hit cannot mask a share that has gone away. The PVR branches in the same method already returned `FailedToExecute` on failure; this brings the generic path into line.
- `schema/version.txt` — `13.15.0`.

## How has this been tested?

Against a library of 1156 movies and 43 TV shows on an SMB share:

```
VideoLibrary.GetMovieDetails  movieid 999999   -> -32098 Not found.
AudioLibrary.GetAlbumDetails  albumid 999999   -> -32098 Not found.
AudioLibrary.GetSongDetails   songid  999999   -> -32098 Not found.
AudioLibrary.GetArtistDetails artistid 999999  -> -32098 Not found.
Player.Open   file that does not exist         -> -32097 Requested item is unavailable.
AudioLibrary.GetAlbums  sort.method "bogus"    -> -32602 Invalid params.  (error.data intact)
Player.Open   movieid 1 (present)              -> OK, plays
```

The last two are the regression checks: genuinely malformed parameters still return `-32602` with their schema-validation `data` payload, and reachable content still plays.

## What is the effect on users?

No visible change in the UI. API clients that match on `-32602` to detect a missing object will need to match `-32098` instead. Clients calling `Player.Open` will now receive an error where they previously received `ACK` for content that never played, which is the point of the change.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [X] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

The wiki will be updated with the new error codes if this is accepted.

There are no automated tests for the JSON-RPC layer to extend — `cmake/treedata/common/tests.txt` has no `xbmc/interfaces/json-rpc` entry, and the handlers are not unit testable in place as they depend on `CServiceBroker`, a database and the application messenger. The verification above was therefore done by hand against a running instance. Happy to discuss what a test approach for this layer should look like separately.

enen92 commented on xbmc/interfaces/json-rpc/VideoLibrary.cpp:
```diff
@@ -95,7 +95,7 @@ JSONRPC_STATUS CVideoLibrary::GetMovieDetails(const std::string &method, ITransp
                                   RequiresAdditionalDetails(MediaTypeMovie, parameterObject)) ||
       infos.m_iDbId <= 0)
-    return InvalidParams;
+    return InvalidObject; //Movie not found
```
mind adding a space after the comment marks?

enen92 reviewed (commented):
I personally have nothing wrong with this but the json interface version needs a bump: https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/version.txt

Unsure if we're still in time for it considering the omega timeline. I guess that since we're still in beta and decided to have another beta it should be fine

malard commented:
i've changed the version to 13.4.1 as i don't believe this is a bug fix nor a major change (so treating it as a minor)
there were more cases of the comments not being to the same standard (so to speak) so touched those also

malard commented:
unless anyone objects I will merge this tomorrow

ksooo reviewed (dismissed):
It seems the new error code is used solely for "not found in database" errors, and even the message contained in the error response is hard-coded for that exact case, thus the new code should be named after that, for example `NotFound` or even `NotFoundInDataBase`.

garbear commented:
This seems like a good feature. Any desire to rebase and get this merged?

garbear commented:
@malard Can this be rebased for inclusion in Beta 2?

malard commented:
I will do it tonight, I could probably get Claude to do a pass on this with supervision to fix up more gaps if you want?

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Garrett Brown ***@***.***>
Sent: Thursday, 16 July 2026 13:34:52
To: xbmc/xbmc ***@***.***>
Cc: Martin Ellis ***@***.***>; Mention ***@***.***>
Subject: Re: [xbmc/xbmc] JSONRPC return custom error -32101 when an object is not found rathe… (PR #24555)

[https://avatars.githubusercontent.com/u/531482?s=20&v=4]garbear left a comment (xbmc/xbmc#24555)<https://github.com/xbmc/xbmc/pull/24555#issuecomment-4991866256>

@malard<https://github.com/malard> Can this be rebased for inclusion in Beta 2?

—
Reply to this email directly, view it on GitHub<https://github.com/xbmc/xbmc/pull/24555?email_source=notifications&email_token=AAEFI55V55E6W5YLCW72XM35FDDWZA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIOJZGE4DMNRSGU3KM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#issuecomment-4991866256>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAEFI5ZVHPI7KAGFFL4TU6L5FDDWZAVCNFSNUABDKJSXA33TNF2G64TZHMYTEMJXGYYTIO2JONZXKZJ3GIYDSMRVGU2DKNZVUF3AE>.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS<https://github.com/notifications/mobile/ios/AAEFI5ZIQHQUVLPYHCWLO4D5FDDWZA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIOJZGE4DMNRSGU3KM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJKTGN5XXIZLSL5UW64Y> and Android<https://github.com/notifications/mobile/android/AAEFI5YTZZQYUIFHRN6GICT5FDDWZA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIOJZGE4DMNRSGU3KM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLTGN5XXIZLSL5QW4ZDSN5UWI>. Download it today!
You are receiving this because you were mentioned.Message ID: ***@***.***>

garbear commented:
Sure, that would be welcome. @kodiai please review

garbear commented:
Is this wanted for Beta 2? We got about a week until release.

garbear commented:
Last chance, do we bump to v23?

malard commented:
I am sorry, I don't have time to contribute right now

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Garrett Brown ***@***.***>
Sent: Saturday, 01 August 2026 18:06:52
To: xbmc/xbmc ***@***.***>
Cc: Martin Ellis ***@***.***>; Mention ***@***.***>
Subject: Re: [xbmc/xbmc] JSONRPC return custom error -32101 when an object is not found rathe… (PR #24555)

[https://avatars.githubusercontent.com/u/531482?s=20&v=4]garbear left a comment (xbmc/xbmc#24555)<https://github.com/xbmc/xbmc/pull/24555#issuecomment-5152467782>

Last chance, do we bump to v23?

—
Reply to this email directly, view it on GitHub<https://github.com/xbmc/xbmc/pull/24555?email_source=notifications&email_token=AAEFI53WVAPWQP7LSN4OAKT5HYPSZA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTKMJVGI2DMNZXHAZKM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#issuecomment-5152467782>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAEFI5675PXXLRSIVVPNE2D5HYPSZAVCNFSNUABDKJSXA33TNF2G64TZHMYTEMJXGYYTIO2JONZXKZJ3GIYDSMRVGU2DKNZVUF3AE>.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS<https://github.com/notifications/mobile/ios/AAEFI52NQGTBPWHP2ELEQ5L5HYPSZA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTKMJVGI2DMNZXHAZKM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJKTGN5XXIZLSL5UW64Y> and Android<https://github.com/notifications/mobile/android/AAEFI56JP34XSKM7TBMFQOD5HYPSZA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTKMJVGI2DMNZXHAZKM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLTGN5XXIZLSL5QW4ZDSN5UWI>. Download it today!
You are receiving this because you were mentioned.Message ID: ***@***.***>

garbear commented:
We understand. If the situation changes, we can retarget this to v22 RC 1, which will be the cuttof for any API bumps.

malard committed:
changed: distinguish not-found and unavailable from invalid params in JSON-RPC

JSON-RPC returned InvalidParams both when parameters were genuinely malformed
and when they were well formed but referenced an object that does not exist.
A client could not tell a bad request from a stale reference.

Player.Open compounded this. Playback is posted asynchronously, so the method
returned ACK even when the item could not be reached, reporting success for
something it never did.

Two statuses are added within the implementation-defined range:

  NotFound    (-32098)  the referenced object does not exist
  Unavailable (-32097)  the reference is valid but the content cannot be reached

The distinction matters because the remedies are opposite: NotFound means the
reference should be discarded, Unavailable means it should be kept and retried
once the storage is back.

VideoLibrary and AudioLibrary now return NotFound where a library object is
absent. Player.Open verifies a single addressable item before posting playback
and returns Unavailable when it cannot be reached; folders, plugins, UPnP,
internet streams and Bluray paths are left to the player as before, and the
directory cache is bypassed so a cached hit cannot mask a share that has gone
away.

The reserved range comment is corrected to -32100..-32000 to match the codes
already in use, so no existing status has to move.

malard commented:
Rebased onto current master and reworked. This is now a single commit, +62/-31, and mergeable.

@ksooo — your review was right, and it changed the shape of this. `InvalidObject` was named and worded for the database case only, so I have dropped it in favour of a general **`NotFound` (-32098)** that also covers the VFS, PVR and add-ons rather than needing a near-duplicate code later. The hardcoded "Not Found in database" message is gone.

While reworking it I found the same problem in a worse form. `Player.Open` posts playback asynchronously and returns `ACK` before anything is attempted, so it reports success even when the file cannot be reached — a client gets "OK" and nothing plays. So there is a second status, **`Unavailable` (-32097)**, for a reference that is valid but whose content is unreachable. The two are deliberately distinct because the remedies are opposite: `NotFound` means discard the reference, `Unavailable` means keep it and retry when the storage returns. The PVR branches of `Player.Open` already returned `FailedToExecute` on failure; this brings the generic path into line with them.

Also picked up along the way: `AudioLibrary` had the same eight artist/album/song cases as `VideoLibrary` and was missed originally, so music and video now behave the same.

One deliberate choice worth flagging: the reserved-range comment is corrected from `-32099..-32000` to `-32100..-32000`. That is one code point more than JSON-RPC 2.0 reserves, but it matches what Kodi already does — `FailedToExecute` has been `-32100` for years — and it means no existing status has to move.

@garbear — you asked twice about getting this in and I did not have the time then. It is ready now if you would like to take a look. Verified by hand against a 1156 movie / 43 show library; the exact calls and responses are in the description, including the regression checks that malformed params still return `-32602` with their `error.data` intact and that reachable content still plays.

malard committed:
changed: apply clang-format to the touched JSON-RPC regions

The PR build job reported formatting deviations in the two files whose edits
land inside legacy-formatted regions.

PlayerOperations.cpp is the reachability guard alone, refilled to the 100
column limit.

JSONRPCUtils.h is largely collateral: the edited lines sit inside
namespace JSONRPC, so clang-format de-indents the enclosing constructs to
match NamespaceIndentation: None, drops the vertical alignment in
OperationPermission, breaks the MethodCall typedef one parameter per line and
rewraps the two permission masks. The doc comment bodies are re-aligned by
hand, as ReflowComments is off and clang-format leaves them at the indent
their opening delimiter no longer has.

The remainder of the header is left as it is, per the guideline against
reformatting preexisting code that a change does not touch.
